File: csharp/src/Noie.Nds.Api/Adapter/V3SyncV1Adapter.cs
using Nds.Sync.V1;

namespace Noie.Nds.Api.Adapter;

/// <summary>
/// [Index: NDS-CSHARP-V3-SYNCV1ADAPTER-000] v3 sync (v1 package) adapter.
///
/// <para>
/// [Semantic] Minimal helpers for v3+ streaming primitives (ResumeToken).
/// </para>
/// </summary>
public static class V3SyncV1Adapter
{
    /// <summary>
    /// [Index: NDS-CSHARP-V3-SYNCV1ADAPTER-010] Create a <see cref="ResumeToken"/> from opaque bytes.
    /// </summary>
    /// <exception cref="ArgumentException">If <paramref name="value"/> is null or empty.</exception>
    public static ResumeToken CreateResumeToken(byte[] value)
    {
        if (value is not { Length: > 0 })
            throw new ArgumentException("value must be non-empty", nameof(value));

        return new ResumeToken { Value = Google.Protobuf.ByteString.CopyFrom(value) };
    }

    /// <summary>
    /// [Index: NDS-CSHARP-V3-SYNCV1ADAPTER-011] Extract opaque bytes from a <see cref="ResumeToken"/>.
    /// </summary>
    public static byte[] ToBytes(ResumeToken token)
    {
        ArgumentNullException.ThrowIfNull(token);
        return token.Value.ToByteArray();
    }
}

File: java/src/main/java/noie/linmimeng/noiedigitalsystem/api/adapter/V3SyncV1Adapter.java
package noie.linmimeng.noiedigitalsystem.api.adapter;

import java.util.Objects;
import noie.linmimeng.noiedigitalsystem.api.proto.sync.v1.ResumeToken;

/**
 * v3 Sync (v1 package) Adapter - minimal helpers for streaming primitives.
 *
 * @since 3.0.0
 */
public final class V3SyncV1Adapter {

    private V3SyncV1Adapter() {
        // [Index] NDS-JAVA-V3-SYNCV1-001 [Constraint] Utility class; instantiation is prohibited.
    }

    /**
     * Create ResumeToken from opaque bytes.
     *
     * @param value opaque bytes (must be non-empty)
     * @return ResumeToken
     * @throws IllegalArgumentException if value is null/empty
     * @since 3.0.0
     */
    public static ResumeToken createResumeToken(byte[] value) {
        if (value == null || value.length == 0) {
            throw new IllegalArgumentException("value must be non-empty");
        }
        return ResumeToken.newBuilder()
            .setValue(com.google.protobuf.ByteString.copyFrom(value))
            .build();
    }

    /**
     * Extract opaque bytes from ResumeToken.
     *
     * @param token ResumeToken (non-null)
     * @return bytes
     * @since 3.0.0
     */
    public static byte[] toBytes(ResumeToken token) {
        Objects.requireNonNull(token, "token");
        return token.getValue().toByteArray();
    }
}

